rust/ql/lib/codeql/rust/internal/PathResolution.qll

```diff
@@ -1563,6 +1666,41 @@ module TraitIsVisible<relevantTraitVisibleSig/2 relevantTraitVisible> {
   }
 }
 
+pragma[nomagic]
+private predicate isDollarCrateSupportedMacroExpansion(Path macroDefPath, AstNode expansion) {
+  exists(MacroCall mc |
+    expansion = mc.getMacroCallExpansion() and
+    macroDefPath = mc.getPath()
+  )
+  or
+  exists(ItemNode adt |
+    expansion = adt.(Adt).getDeriveMacroExpansion(_) and
+    macroDefPath = adt.getAttr("derive").getMeta().getPath()
+  )
+}
+
+pragma[nomagic]
+predicate isInDollarCrateSupportedMacroExpansion(File macroDefFile, AstNode n) {
+  exists(Path macroDefPath |
+    isDollarCrateSupportedMacroExpansion(macroDefPath, n) and
+    macroDefFile = resolvePathCand(macroDefPath).getFile()
+  )
+  or
+  isInDollarCrateSupportedMacroExpansion(macroDefFile, n.getParentNode())
```

Comment:
This predicate holds for all AST nodes within an expansion, but we only care about `$crate` notes. And for expansions within expansions we end up with `2 * nrOfAstNodes` tuples?

Wouldn't it have been more efficient to have

1. a predicate from `$crate` paths to their enclosing expansion and
2. a predicate from a macro call expansion to its enclosing macro call expansion.

And then use these in `resolveDollarCrate` to find all the potential crates? That would avoid a predicate that holds for all AST nodes within expansions.

Reply:
Thanks, after looking up what `doublyBoundedFastTC` I think that approach is really clean!